docs/pipeline_semantics.md

```diff
@@ -0,0 +1,172 @@
+  messages, instead, it waits for a message to be sent to all downstream nodes before processing the next message (see
+  [backpressure](#Backpressure)). A fan-out node is automatically created for all pipelines.
+- **Destination processors** - these processors receive only messages that are meant to be sent to a specific
+  destination connector. Destination processors are created by specifying the corresponding destination connector as the
+  parent entity. Processors are not required for starting a pipeline.
+- **Destination connectors** - represents the code that communicates with a 3rd party system and continuously receives
+  records from Conduit and writes them to the destination (e.g.
+  [kafka connector](https://github.com/conduitio/conduit-connector-kafka)). Every destination connector is managed by a
+  destination node that receives records and sends them to the connector. A pipeline requires at least one destination
+  connector to be runnable.
+
+There are additional internal nodes that Conduit adds to each pipeline not shown in the diagram, as they are
+inconsequential for the purpose of this document (e.g. nodes for gathering metrics, nodes for managing acknowledgments,
+etc.).
+
+## Message
+
+A message is a wrapper around a record that manages the record's lifecycle as it flows through the pipeline. Messages
+are created in source nodes when they receive records from the source connector, and they are passed down the pipeline
+between nodes until they are acked or nacked. Nodes are only allowed to hold a reference to a single message at a time,
+meaning that they need to pass the message to the next node before taking another message¹. This also means there is no
+explicit buffer in Conduit, a pipeline can only hold only as many messages as there are nodes in the pipeline (see
+[backpressure](#Backpressure) for more information).
+
+¹This might change in the future if we decide to add support for multi-message transforms.
+
+### Message states
+
+A message can be in one of 3 states:
+
+- **Open** - all messages start in the open state. This means the message is currently in processing, either by a node
+  or a destination connector. A pipeline won't stop until all messages transition from the open state into one of the
+  other two states.
+- **Acked** - a message was successfully processed and acknowledged. This can be done either by a processor (e.g. it
+  filtered the message out) or by a destination. If a pipeline contains multiple destinations, the message needs to be
+  acknowledged by all destinations before it is marked as acked. Acks are propagated back to the source connector and
+  can be used to advance the position in the source system if applicable.
+- **Nacked** - a message was negatively acknowledged either by a processor (e.g. a transform failed) or by a
+  destination. If a pipeline contains multiple destinations, the message needs to be negatively acknowledged by at least
```

Comment:
> NACK (received with errors or was otherwise unreadable, etc)

I assume you're talking about the NACK signal in the context of network protocols? While the term is borrowed from that space, it does not necessarily mean the same thing in Conduit. Here a nack means that the processing of a message failed and resulted in an error.

I'll try to clarify this part!